src/hardware/voodoo.cpp

```diff
@@ -0,0 +1,7175 @@
+	{
+		if (exponent > -32)
+			result >>= -exponent;
+		else
+			result = 0;
+	}
+	else
+	{
+		if (exponent < 32)
+			result <<= exponent;
+		else
+			result = 0x7fffffff;
+	}
+	if (data & 0x80000000)
+		result = -result;
+	return result;
+}
+
+inline INT64 float_to_int64(UINT32 data, int fixedbits)
+{
+	int exponent = ((data >> 23) & 0xff) - 127 - 23 + fixedbits;
+	INT64 result = (data & 0x7fffff) | 0x800000;
+	if (exponent < 0)
+	{
+		if (exponent > -64)
+			result >>= -exponent;
+		else
+			result = 0;
+	}
+	else
+	{
+		if (exponent < 64)
+			result <<= exponent;
+		else
+			result = LONGTYPE(0x7fffffffffffffff);
+	}
+	if (data & 0x80000000)
+		result = -result;
+	return result;
+}
+
+
+
+#ifdef C_ENABLE_VOODOO_OPENGL
+/*************************************
+ *
+ *  Rasterizer inlines
+ *
+ *************************************/
+
+inline UINT32 normalize_color_path(UINT32 eff_color_path)
+{
+	/* ignore the subpixel adjust and texture enable flags */
+	eff_color_path &= ~((1 << 26) | (1 << 27));
+
+	return eff_color_path;
+}
+
+inline UINT32 normalize_alpha_mode(UINT32 eff_alpha_mode)
+{
+	/* always ignore alpha ref value */
+	eff_alpha_mode &= ~(0xff << 24);
+
+	/* if not doing alpha testing, ignore the alpha function and ref value */
+	if (!ALPHAMODE_ALPHATEST(eff_alpha_mode))
+		eff_alpha_mode &= ~(7 << 1);
+
+	/* if not doing alpha blending, ignore the source and dest blending factors */
+	if (!ALPHAMODE_ALPHABLEND(eff_alpha_mode))
+		eff_alpha_mode &= ~((15 << 8) | (15 << 12) | (15 << 16) | (15 << 20));
+
+	return eff_alpha_mode;
+}
+
+inline UINT32 normalize_fog_mode(UINT32 eff_fog_mode)
+{
+	/* if not doing fogging, ignore all the other fog bits */
+	if (!FOGMODE_ENABLE_FOG(eff_fog_mode))
+		eff_fog_mode = 0;
+
+	return eff_fog_mode;
+}
+
+inline UINT32 normalize_fbz_mode(UINT32 eff_fbz_mode)
+{
+	/* ignore the draw buffer */
+	eff_fbz_mode &= ~(3 << 14);
+
+	return eff_fbz_mode;
+}
+
+inline UINT32 normalize_tex_mode(UINT32 eff_tex_mode)
+{
+	/* ignore the NCC table and seq_8_downld flags */
+	eff_tex_mode &= ~((1 << 5) | (1 << 31));
+
+	/* classify texture formats into 3 format categories */
+	if (TEXMODE_FORMAT(eff_tex_mode) < 8)
+		eff_tex_mode = (eff_tex_mode & ~(0xf << 8)) | (0 << 8);
+	else if (TEXMODE_FORMAT(eff_tex_mode) >= 10 && TEXMODE_FORMAT(eff_tex_mode) <= 12)
+		eff_tex_mode = (eff_tex_mode & ~(0xf << 8)) | (10 << 8);
+	else
+		eff_tex_mode = (eff_tex_mode & ~(0xf << 8)) | (8 << 8);
+
+	return eff_tex_mode;
+}
+
+inline UINT32 compute_raster_hash(const raster_info* info)
+{
+	UINT32 hash;
+
+	/* make a hash */
+	hash = info->eff_color_path;
+	hash = (hash << 1) | (hash >> 31);
+	hash ^= info->eff_fbz_mode;
+	hash = (hash << 1) | (hash >> 31);
+	hash ^= info->eff_alpha_mode;
+	hash = (hash << 1) | (hash >> 31);
+	hash ^= info->eff_fog_mode;
+	hash = (hash << 1) | (hash >> 31);
+	hash ^= info->eff_tex_mode_0;
+	hash = (hash << 1) | (hash >> 31);
+	hash ^= info->eff_tex_mode_1;
+
+	return hash % RASTER_HASH_SIZE;
+}
```

Comment:
What's this hash?

Reply:
It's for storing "rasterizers" in C-implemented hash table. Used in OpenGL mode only though...

Not sure if we need it; as for now it's ifdefed-out behind `C_ENABLE_VOODOO_OPENGL` so not touching it.